### ISSUE TITLE
Fails to compile with nvidia-cuda-toolkit-12.4.0

### ISSUE THREAD

**Issue by blshkv** [pytorch/pytorch#122169]
### 🐛 Describe the bug

it fails to compile with nvidia-cuda-toolkit-12.4.0 (compiles fine with 12.3.2)
I tried pytorch both 2.1.2 and 2.2.1
https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/cuda/LinearAlgebra.cu#L131

```
/data/notmpfs/portage/sci-libs/caffe2-2.2.1-r1/work/pytorch-2.2.1/aten/src/ATen/native/cuda/LinearAlgebra.cu(129): error: namespace "thrust" has no member "swap"
        thrust::swap(
                ^
1 error detected in the compilation of "/data/notmpfs/portage/sci-libs/caffe2-2.2.1-r1/work/pytorch-2.2.1/aten/src/ATen/native/cuda/LinearAlgebra.cu".
```

### Versions

```
Collecting environment information...
PyTorch version: 2.1.0a0
Is debug build: False
CUDA used to build PyTorch: 12.3
ROCM used to build PyTorch: N/A

OS: Gentoo Linux (x86_64)
GCC version: (Gentoo Hardened 12.3.1_p20240209 p3) 12.3.1 20240209
Clang version: 17.0.6
CMake version: version 3.27.9
Libc version: glibc-2.38

Python version: 3.12.2 (main, Feb 24 2024, 19:52:20) [GCC 12.3.1 20240112] (64-bit runtime)
Python platform: Linux-6.6.21-gentoo-x86_64-Intel-R-_Core-TM-_i7-10750H_CPU_@_2.60GHz-with-glibc2.38
Is CUDA available: True
CUDA runtime version: 12.3.107
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce GTX 1650
Nvidia driver version: 535.161.07
cuDNN version: Probably one of the following:
/opt/cuda/targets/x86_64-linux/lib/libcudnn.so.8.8.0
/opt/cuda/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.8.0
/opt/cuda/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.8.0
/opt/cuda/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.8.0
/opt/cuda/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.8.0
/opt/cuda/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.8.0
/opt/cuda/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.8.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      39 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             12
On-line CPU(s) list:                0-11
Vendor ID:                          GenuineIntel
Model name:                         Intel(R) Core(TM) i7-10750H CPU @ 2.60GHz
CPU family:                         6
Model:                              165
Thread(s) per core:                 2
Core(s) per socket:                 6
Socket(s):                          1
Stepping:                           2
CPU(s) scaling MHz:                 60%
CPU max MHz:                        5000.0000
CPU min MHz:                        800.0000
BogoMIPS:                           5199.98
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid mpx rdseed adx smap clflushopt intel_pt xsaveopt xsavec xgetbv1 xsaves dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp vnmi pku ospke md_clear flush_l1d arch_capabilities
Virtualization:                     VT-x
L1d cache:                          192 KiB (6 instances)
L1i cache:                          192 KiB (6 instances)
L2 cache:                           1.5 MiB (6 instances)
L3 cache:                           12 MiB (1 instance)
Vulnerability Gather data sampling: Vulnerable: No microcode
Vulnerability Itlb multihit:        KVM: Mitigation: VMX disabled
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Vulnerable: Clear CPU buffers attempted, no microcode; SMT vulnerable
Vulnerability Retbleed:             Mitigation; Enhanced IBRS
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Vulnerable: eIBRS with unprivileged eBPF
Vulnerability Srbds:                Vulnerable: No microcode
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] torch==2.1.2
[conda] Could not collect
```

cc @malfet @seemethere @ptrblck

**Comment by christian-heusel** [comment]
Shouldnt this be fixed with https://github.com/pytorch/pytorch/commit/2a440348958b3f0a2b09458bd76fe5959b371c0c? I'm hitting the same error right now on 2.2.0

**Comment by blshkv** [comment]
ok, it fixed the issue. The next one:
```
/data/notmpfs/portage/sci-libs/caffe2-2.2.1-r1/work/pytorch-2.2.1/aten/src/ATen/core/IListRef_inl.h:171:33: note: the temporary was destroyed at the end of the full expression ‘(& it)->c10::impl::ListIterator<std::optional<at::Tensor>, __gnu_cxx::__normal_iterator<c10::IValue*, std::vector<c10::IValue> > >::operator*().c10::impl::ListElementReference<std::optional<at::Tensor>, __gnu_cxx::__normal_iterator<c10::IValue*, std::vector<c10::IValue> > >::get()’
  171 |     const auto& ivalue = (*it).get();
      |                      ~~~~~~~~~~~^~
/data/notmpfs/portage/sci-libs/caffe2-2.2.1-r1/work/pytorch-2.2.1/aten/src/ATen/core/boxing/impl/boxing.h: At global scope:
/data/notmpfs/portage/sci-libs/caffe2-2.2.1-r1/work/pytorch-2.2.1/aten/src/ATen/core/boxing/impl/boxing.h:41:105: error: expected primary-expression before ‘>’ token
   41 | struct has_ivalue_to<T, guts::void_t<decltype(std::declval<IValue>().to<T>())>>
      |                                                                                                         ^
/data/notmpfs/portage/sci-libs/caffe2-2.2.1-r1/work/pytorch-2.2.1/aten/src/ATen/core/boxing/impl/boxing.h:41:108: error: expected primary-expression before ‘)’ token
   41 | struct has_ivalue_to<T, guts::void_t<decltype(std::declval<IValue>().to<T>())>>

```

**Comment by christian-heusel** [comment]
Seems like your facing the same issues like I do (currently rebuilding the pytorch package for archlinux) 😆 
Not verified to work yet, but I'll try to backport c3780010a58a84920335296ee5f091a0db18259f 3baade4425f7234bdbb31a9d3283c39cb7d69999 🤷🏻

**Comment by blshkv** [comment]
hm... thanks for pointing it all out... It looks like it's better to wait for a next release

**Comment by colesbury** [comment]
Closing the issue because it sounds like it's fixed in `main`.

**Comment by colesbury** [comment]
Please re-open the issue if that's not the case.

**Comment by christian-heusel** [comment]
So the second issue that was described here is not yet fixed and is apparently a compiler bug in nvcc.. See the fix here (credits go to @lahwaacz): https://gitlab.archlinux.org/archlinux/packaging/packages/python-pytorch/-/blob/main/python-pytorch-fix-cuda-12_4.patch?ref_type=heads

**Comment by blshkv** [comment]
@colesbury I'm unable to re-open the bug, but you might want to fix the last bug mentioned by Christian. BTW, thanks Christian!

**Comment by mfbalin** [comment]
> ok, it fixed the issue. The next one:
> 
> ```
> /data/notmpfs/portage/sci-libs/caffe2-2.2.1-r1/work/pytorch-2.2.1/aten/src/ATen/core/IListRef_inl.h:171:33: note: the temporary was destroyed at the end of the full expression ‘(& it)->c10::impl::ListIterator<std::optional<at::Tensor>, __gnu_cxx::__normal_iterator<c10::IValue*, std::vector<c10::IValue> > >::operator*().c10::impl::ListElementReference<std::optional<at::Tensor>, __gnu_cxx::__normal_iterator<c10::IValue*, std::vector<c10::IValue> > >::get()’
>   171 |     const auto& ivalue = (*it).get();
>       |                      ~~~~~~~~~~~^~
> /data/notmpfs/portage/sci-libs/caffe2-2.2.1-r1/work/pytorch-2.2.1/aten/src/ATen/core/boxing/impl/boxing.h: At global scope:
> /data/notmpfs/portage/sci-libs/caffe2-2.2.1-r1/work/pytorch-2.2.1/aten/src/ATen/core/boxing/impl/boxing.h:41:105: error: expected primary-expression before ‘>’ token
>    41 | struct has_ivalue_to<T, guts::void_t<decltype(std::declval<IValue>().to<T>())>>
>       |                                                                                                         ^
> /data/notmpfs/portage/sci-libs/caffe2-2.2.1-r1/work/pytorch-2.2.1/aten/src/ATen/core/boxing/impl/boxing.h:41:108: error: expected primary-expression before ‘)’ token
>    41 | struct has_ivalue_to<T, guts::void_t<decltype(std::declval<IValue>().to<T>())>>
> ```

I have this compiler error with GCC 13.2.

The error output below:

```
/home/mfbalin/.venvs/venv/lib/python3.12/site-packages/torch/include/ATen/core/boxing/impl/boxing.h:42:103: error: expected primary-expression before ‘>’ token
   42 | struct has_ivalue_to<T, std::void_t<decltype(std::declval<IValue>().to<T>())>>
      |                                                                                                       ^
/home/mfbalin/.venvs/venv/lib/python3.12/site-packages/torch/include/ATen/core/boxing/impl/boxing.h:42:106: error: expected primary-expression before ‘)’ token
   42 | struct has_ivalue_to<T, std::void_t<decltype(std::declval<IValue>().to<T>())>>
      |                                            
```

**Comment by Xynonners** [comment]
```
/vol/zraid1/Projects/AI/Self/variations/venv/lib/python3.11/site-packages/torch/include/ATen/core/boxing/impl/boxing.h:42:103: error: expected primary-expression before ‘>’ token
       42 | struct has_ivalue_to<T, std::void_t<decltype(std::declval<IValue>().to<T>())>>
```
same issue here.

**Comment by tavyscrolls** [comment]
Problem persists in torch 2.4.0 torchvision 0.19.0 Debian GNU 13 solved by upgrading to nightly, building torch, upgrading torch 2.3 to 2.4, then building torchvision. It is possible I missed a flag somewhere and the extra steps are redundant. Compile flags set:

<summary>
<details>
Compile flags

- GLIBCXX_USE_CXX11_ABI=1
- USE_CUDNN=1
- USE_CUDA=1
- USE_FBGEMM=1
- USE_NUMPY=1
- USE_QNNPACK=1
- USE_TENSORPIPE=1
- USE_GLOO=0
- USE_MPI=0
- USE_SYSTEM_NCCL=1
- USE_IBVERBS=0
- USE_OPENCV=1
- USE_OPENMP=1
- USE_FFMPEG=1
- USE_FLASH_ATTENTION=1
- USE_MEM_EFF_ATTENTION=1
- USE_LEVELDB=1
- USE_LMDB=1

</details>
</summary>

I also set my cuda 12.4 path and gpu architecture explicitly. Hope this helps.

**Comment by znshje** [comment]
Same issue. PyTorch 2.3.0 + CUDA 12.5

```
/home/a/anaconda3/envs/a/lib/python3.8/site-packages/torch/include/ATen/core/boxing/impl/boxing.h:42:106: 错误：expected primary-expression before ‘)’ token
   42 | struct has_ivalue_to<T, std::void_t<decltype(std::declval<IValue>().to<T>())>>
      |                                                                                                          ^
```

**Comment by znshje** [comment]
I fixed the issue by adding an extra NVCC compiler arg `-std=c++20`, and the `setup.py` looks like:

```python
from setuptools import setup
from torch.utils.cpp_extension import CUDAExtension, BuildExtension

setup(
    name='some_lib',  # used by `pip install`
    version='0.0.1',
    description='',
    cmdclass={
        'build_ext': BuildExtension
    },
    ext_modules=[
        CUDAExtension(
            'some_lib',
            [
                'src/aaa.cu',
            ],
            extra_compile_args={'cxx': ['-g'],
                                'nvcc': ['-O2', '-allow-unsupported-compiler', '-std=c++20']})
    ],
    setup_requires=["pybind11"],
    install_requires=["pybind11"],
    python_requires='>=3.8',
    include_package_data=True,
    zip_safe=False,
)
```

Hope this helps.

**Comment by zhanwenchen** [comment]
Another way is to replace this block in [`/aten/src/ATen/core/boxing/impl/boxing.h`](https://github.com/pytorch/pytorch/blob/v2.3.0/aten/src/ATen/core/boxing/impl/boxing.h#L36C1-L48C3) from the v2.3.0 branch:

```python
// has_ivalue_to<T> tests the presence/absence of instance method IValue::to<T>()
//
template <class T, class Enable = void>
struct has_ivalue_to : std::false_type {};

template <class T>
struct has_ivalue_to<T, std::void_t<decltype(std::declval<IValue>().to<T>())>>
: std::true_type
{};

//
// boxing predicates
//
```

to the nightly version:

```python
// has_ivalue_to<T> tests the presence/absence of instance method IValue::to<T>()
//
template <class T, class Enable = void>
struct has_ivalue_to : std::false_type {};

template <class T>
struct ivalue_to_helper
{
    using type = decltype(std::declval<IValue>().template to<T>());
};
template <class T>
using ivalue_to_helper_t = typename ivalue_to_helper<T>::type;

template <class T>
struct has_ivalue_to<T, std::void_t<ivalue_to_helper_t<T>>>
: std::true_type
{};

//
// boxing predicates
//
```

**Comment by zhanwenchen** [comment]
NOTE: this has not been fixed in today's 2.3.1 release: https://github.com/pytorch/pytorch/blob/v2.3.1/aten/src/ATen/core/boxing/impl/boxing.h#L36-L48

**Comment by zhanwenchen** [comment]
This has been fixed in the v2.4.0 release: https://github.com/pytorch/pytorch/blob/v2.4.0/aten/src/ATen/core/boxing/impl/boxing.h#L36-L56
